File: Interview Preparation Kit - CPP/01. Warm-up Challenges/002. Counting Valleys.cpp
// Problem: https://www.hackerrank.com/challenges/counting-valleys/problem
// Score: 15

#include <iostream>
#include <string>
using namespace std;

int main() {
    int n;
    cin >> n;

    string str;
    cin >> str;

    int ans = 0;

    int current_level = 0;

    for (char c: str){
        if (c == 'D') {
            if (current_level == 0){
                ans++;
            }

            current_level--;
        }

        else if (c == 'U') current_level++;
    }

    cout << ans;

    return 0;
}

File: Interview Preparation Kit - Python/01. Warm-up Challenges/002. Counting Valleys.py
# Problem: https://www.hackerrank.com/challenges/counting-valleys/problem
#  Score: 15


def counting_valleys(n, s):
    current_level = 0
    count = 0
    for i in range(n):
        if s[i] == 'U':
            current_level += 1
        elif s[i] == 'D':
            current_level -= 1
            if current_level == -1:
                count += 1
    return count


n = int(input().strip())
s = input().strip()
result = counting_valleys(n, s)
print(result)
